# SF Salaries Data Exploration with Pandas

In [1]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All" 
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

/kaggle/input/sf-salaries/Salaries.csv
/kaggle/input/sf-salaries/database.sqlite


In [2]:
sal = pd.read_csv('/kaggle/input/sf-salaries/Salaries.csv')


# the for loop below is to ensure consistency on the datatype of values provided in columns; BasePay, OvertimePay, OtherPay and Benefits. 

for col in sal[['BasePay','OvertimePay','OtherPay','Benefits']].select_dtypes(include=['object']).columns:
    sal[col] = pd.to_numeric(sal[col], errors = 'coerce')

<ipython-input-2-2a12daf4a160>:1: DtypeWarning: Columns (3,4,5,6,12) have mixed types. Specify dtype option on import or set low_memory=False.
  sal = pd.read_csv('/kaggle/input/sf-salaries/Salaries.csv')


In [3]:
sal.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 148654 entries, 0 to 148653
Data columns (total 13 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   Id                148654 non-null  int64  
 1   EmployeeName      148654 non-null  object 
 2   JobTitle          148654 non-null  object 
 3   BasePay           148045 non-null  float64
 4   OvertimePay       148650 non-null  float64
 5   OtherPay          148650 non-null  float64
 6   Benefits          112491 non-null  float64
 7   TotalPay          148654 non-null  float64
 8   TotalPayBenefits  148654 non-null  float64
 9   Year              148654 non-null  int64  
 10  Notes             0 non-null       float64
 11  Agency            148654 non-null  object 
 12  Status            38119 non-null   object 
dtypes: float64(7), int64(2), object(4)
memory usage: 14.7+ MB


In [4]:
sal.head(3)

Id    EmployeeName                                        JobTitle  \
0   1  NATHANIEL FORD  GENERAL MANAGER-METROPOLITAN TRANSIT AUTHORITY   
1   2    GARY JIMENEZ                 CAPTAIN III (POLICE DEPARTMENT)   
2   3  ALBERT PARDINI                 CAPTAIN III (POLICE DEPARTMENT)   

     BasePay  OvertimePay   OtherPay  Benefits   TotalPay  TotalPayBenefits  \
0  167411.18         0.00  400184.25       NaN  567595.43         567595.43   
1  155966.02    245131.88  137811.38       NaN  538909.28         538909.28   
2  212739.13    106088.18   16452.60       NaN  335279.91         335279.91   

   Year  Notes         Agency Status  
0  2011    NaN  San Francisco    NaN  
1  2011    NaN  San Francisco    NaN  
2  2011    NaN  San Francisco    NaN

The average BasePay of salaries in this dataset:

In [5]:
sal['BasePay'].mean()

66325.4488404877

** What is the highest amount of OvertimePay in the dataset ? **

In [6]:
sal['OvertimePay'].max()

245131.88

** What is the job title of JOSEPH DRISCOLL ? Note: Use all caps, otherwise you may get an answer that doesn't match up (there is also a lowercase Joseph Driscoll). **

In [7]:
sal[sal['EmployeeName'] == 'JOSEPH DRISCOLL']['JobTitle']

24    CAPTAIN, FIRE SUPPRESSION
Name: JobTitle, dtype: object

** How much does JOSEPH DRISCOLL make (including benefits)? **

In [8]:
sal[sal['EmployeeName'] == 'JOSEPH DRISCOLL']['TotalPayBenefits']

24    270324.91
Name: TotalPayBenefits, dtype: float64

** What is the name of highest paid person (including benefits)?**

In [9]:
sal.loc[[sal['TotalPayBenefits'].argmax()]]

Id    EmployeeName                                        JobTitle  \
0   1  NATHANIEL FORD  GENERAL MANAGER-METROPOLITAN TRANSIT AUTHORITY   

     BasePay  OvertimePay   OtherPay  Benefits   TotalPay  TotalPayBenefits  \
0  167411.18          0.0  400184.25       NaN  567595.43         567595.43   

   Year  Notes         Agency Status  
0  2011    NaN  San Francisco    NaN

** What is the name of lowest paid person (including benefits)? Do you notice something strange about how much he or she is paid?**

In [10]:
sal.loc[[sal['TotalPayBenefits'].argmin()]]

Id EmployeeName                    JobTitle  BasePay  OvertimePay  \
148653  148654    Joe Lopez  Counselor, Log Cabin Ranch      0.0          0.0   

        OtherPay  Benefits  TotalPay  TotalPayBenefits  Year  Notes  \
148653   -618.13       0.0   -618.13           -618.13  2014    NaN   

               Agency Status  
148653  San Francisco     PT

Joe Lopez was the lowest paid employee in San Francisco with a deficit pay of -618.13

** What was the average (mean) BasePay of all employees per year? (2011-2014) ? **

In [11]:
sal.groupby('Year').mean(numeric_only = True)['BasePay']

Year
2011    63595.956517
2012    65436.406857
2013    69630.030216
2014    66564.421924
Name: BasePay, dtype: float64

** How many unique job titles are there? **

In [12]:
sal['JobTitle'].nunique()

2159

** What are the top 5 most common jobs? **

In [13]:
sal['JobTitle'].value_counts().head(5)

JobTitle
Transit Operator                7036
Special Nurse                   4389
Registered Nurse                3736
Public Svc Aide-Public Works    2518
Police Officer 3                2421
Name: count, dtype: int64

** How many Job Titles were represented by only one person in 2013? (e.g. Job Titles with only one occurence in 2013?) **

In [14]:
sal[sal['Year'] == 2013]['JobTitle'].value_counts()[lambda x: x == 1].count()

202

** How many people have the word Chief in their job title? (This is pretty tricky) **

In [15]:
sal['JobTitle'].str.contains(r'\bChief\b', case = False, regex = True).sum()

627

** Bonus: Is there a correlation between length of the Job Title string and Salary? **

In [16]:
sal['JobTitle'].str.len().corr(sal['TotalPayBenefits'])

-0.036878445932606675